{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2adb8871-f454-4832-831f-9a171520d894",
   "metadata": {},
   "source": [
    "### BEstimate Analysis\n",
    "\n",
    "##### gRNA libraries for sickle cell and cystic fibrosis\n",
    "##### Used Parameters\n",
    "- Editing window: 3-9\n",
    "- sgRNA length: 20 nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6c8c1150-4406-4a22-9720-3c0161ca08b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/software/isg/languages/Python-3.9.10/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.4\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "import pandas, numpy, requests\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "data_path = \"/lustre/scratch125/casm/team215mg/cd7/BEstimate/BEstimate/paper/data/\"\n",
    "guide_path = \"/lustre/scratch125/casm/team215mg/cd7/BEstimate/BEstimate/paper/output/examples/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92baa79a-d076-400b-a534-34c6a3d1f9c8",
   "metadata": {},
   "source": [
    "#### *HBB*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "84b34c0c-f332-4795-8216-9ffa61ebdcc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "hbb_wt = pandas.read_csv(\"../output/examples/HBB_ABE_NGN_edit_df.csv\")\n",
    "hbb_mut = pandas.read_csv(\"../output/examples/HBB_ABE_NGN_mut_edit_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7576e5f8-041a-417c-85eb-81174274066a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>CRISPR_PAM_Sequence</th>\n",
       "      <th>gRNA_Target_Sequence</th>\n",
       "      <th>Location</th>\n",
       "      <th>Edit_Location</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Exon_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>gRNA_flanking_sequences</th>\n",
       "      <th>Edit_in_Exon</th>\n",
       "      <th>Edit_in_CDS</th>\n",
       "      <th>GC%</th>\n",
       "      <th># Edits/guide</th>\n",
       "      <th>Poly_T</th>\n",
       "      <th>mutation_on_guide</th>\n",
       "      <th>guide_change_mutation</th>\n",
       "      <th>mutation_on_window</th>\n",
       "      <th>mutation_on_PAM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBB</td>\n",
       "      <td>ACCATTGGAAAAGCAACCCCTGC</td>\n",
       "      <td>ACCATTGGAAAAGCAACCCC</td>\n",
       "      <td>11:5229377-5229399</td>\n",
       "      <td>5229391</td>\n",
       "      <td>left</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HBB</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
       "      <td>11:5229372-5229394</td>\n",
       "      <td>5229391</td>\n",
       "      <td>left</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HBB</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
       "      <td>11:5229372-5229394</td>\n",
       "      <td>5229390</td>\n",
       "      <td>left</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HBB</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
       "      <td>11:5229372-5229394</td>\n",
       "      <td>5229389</td>\n",
       "      <td>left</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol      CRISPR_PAM_Sequence  gRNA_Target_Sequence  \\\n",
       "0         HBB  ACCATTGGAAAAGCAACCCCTGC  ACCATTGGAAAAGCAACCCC   \n",
       "1         HBB  TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
       "2         HBB  TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
       "3         HBB  TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
       "\n",
       "             Location  Edit_Location Direction  Strand          Gene_ID  \\\n",
       "0  11:5229377-5229399        5229391      left      -1  ENSG00000244734   \n",
       "1  11:5229372-5229394        5229391      left      -1  ENSG00000244734   \n",
       "2  11:5229372-5229394        5229390      left      -1  ENSG00000244734   \n",
       "3  11:5229372-5229394        5229389      left      -1  ENSG00000244734   \n",
       "\n",
       "  Transcript_ID Exon_ID  ...  gRNA_flanking_sequences  Edit_in_Exon  \\\n",
       "0           NaN     NaN  ...                      NaN         False   \n",
       "1           NaN     NaN  ...                      NaN         False   \n",
       "2           NaN     NaN  ...                      NaN         False   \n",
       "3           NaN     NaN  ...                      NaN         False   \n",
       "\n",
       "   Edit_in_CDS   GC%  # Edits/guide  Poly_T  mutation_on_guide  \\\n",
       "0        False  50.0              1   False              False   \n",
       "1        False  55.0              4   False              False   \n",
       "2        False  55.0              4   False              False   \n",
       "3        False  55.0              4   False              False   \n",
       "\n",
       "   guide_change_mutation  mutation_on_window  mutation_on_PAM  \n",
       "0                  False               False            False  \n",
       "1                  False               False            False  \n",
       "2                  False               False            False  \n",
       "3                  False               False            False  \n",
       "\n",
       "[4 rows x 21 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hbb_wt[:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a702ced0-6424-400b-98fb-7d5d0a99a95b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>CRISPR_PAM_Sequence</th>\n",
       "      <th>gRNA_Target_Sequence</th>\n",
       "      <th>Location</th>\n",
       "      <th>Edit_Location</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Exon_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>gRNA_flanking_sequences</th>\n",
       "      <th>Edit_in_Exon</th>\n",
       "      <th>Edit_in_CDS</th>\n",
       "      <th>GC%</th>\n",
       "      <th># Edits/guide</th>\n",
       "      <th>Poly_T</th>\n",
       "      <th>mutation_on_guide</th>\n",
       "      <th>guide_change_mutation</th>\n",
       "      <th>mutation_on_window</th>\n",
       "      <th>mutation_on_PAM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBB</td>\n",
       "      <td>ACCATTGGAAAAGCAACCCCTGC</td>\n",
       "      <td>ACCATTGGAAAAGCAACCCC</td>\n",
       "      <td>11:5229377-5229399</td>\n",
       "      <td>5229391</td>\n",
       "      <td>left</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HBB</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
       "      <td>11:5229372-5229394</td>\n",
       "      <td>5229391</td>\n",
       "      <td>left</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HBB</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
       "      <td>11:5229372-5229394</td>\n",
       "      <td>5229390</td>\n",
       "      <td>left</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HBB</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
       "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
       "      <td>11:5229372-5229394</td>\n",
       "      <td>5229389</td>\n",
       "      <td>left</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol      CRISPR_PAM_Sequence  gRNA_Target_Sequence  \\\n",
       "0         HBB  ACCATTGGAAAAGCAACCCCTGC  ACCATTGGAAAAGCAACCCC   \n",
       "1         HBB  TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
       "2         HBB  TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
       "3         HBB  TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
       "\n",
       "             Location  Edit_Location Direction  Strand          Gene_ID  \\\n",
       "0  11:5229377-5229399        5229391      left      -1  ENSG00000244734   \n",
       "1  11:5229372-5229394        5229391      left      -1  ENSG00000244734   \n",
       "2  11:5229372-5229394        5229390      left      -1  ENSG00000244734   \n",
       "3  11:5229372-5229394        5229389      left      -1  ENSG00000244734   \n",
       "\n",
       "  Transcript_ID Exon_ID  ...  gRNA_flanking_sequences  Edit_in_Exon  \\\n",
       "0           NaN     NaN  ...                      NaN         False   \n",
       "1           NaN     NaN  ...                      NaN         False   \n",
       "2           NaN     NaN  ...                      NaN         False   \n",
       "3           NaN     NaN  ...                      NaN         False   \n",
       "\n",
       "   Edit_in_CDS   GC%  # Edits/guide  Poly_T  mutation_on_guide  \\\n",
       "0        False  50.0              1   False              False   \n",
       "1        False  55.0              4   False              False   \n",
       "2        False  55.0              4   False              False   \n",
       "3        False  55.0              4   False              False   \n",
       "\n",
       "   guide_change_mutation  mutation_on_window  mutation_on_PAM  \n",
       "0                  False               False            False  \n",
       "1                  False               False            False  \n",
       "2                  False               False            False  \n",
       "3                  False               False            False  \n",
       "\n",
       "[4 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hbb_mut[:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "104eebf2-ec25-4280-803a-2105a69e485c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>CRISPR_PAM_Sequence</th>\n",
       "      <th>gRNA_Target_Sequence</th>\n",
       "      <th>Location</th>\n",
       "      <th>Edit_Location</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Exon_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>gRNA_flanking_sequences</th>\n",
       "      <th>Edit_in_Exon</th>\n",
       "      <th>Edit_in_CDS</th>\n",
       "      <th>GC%</th>\n",
       "      <th># Edits/guide</th>\n",
       "      <th>Poly_T</th>\n",
       "      <th>mutation_on_guide</th>\n",
       "      <th>guide_change_mutation</th>\n",
       "      <th>mutation_on_window</th>\n",
       "      <th>mutation_on_PAM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2315</th>\n",
       "      <td>HBB</td>\n",
       "      <td>ACTTCTCCACAGGAGTCAGATGC</td>\n",
       "      <td>ACTTCTCCACAGGAGTCAGA</td>\n",
       "      <td>11:5226994-5227016</td>\n",
       "      <td>5227002</td>\n",
       "      <td>right</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSG00000244734</td>\n",
       "      <td>ENST00000335295</td>\n",
       "      <td>ENSE00001829867</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo_Symbol      CRISPR_PAM_Sequence  gRNA_Target_Sequence  \\\n",
       "2315         HBB  ACTTCTCCACAGGAGTCAGATGC  ACTTCTCCACAGGAGTCAGA   \n",
       "\n",
       "                Location  Edit_Location Direction  Strand          Gene_ID  \\\n",
       "2315  11:5226994-5227016        5227002     right      -1  ENSG00000244734   \n",
       "\n",
       "        Transcript_ID          Exon_ID  ...  gRNA_flanking_sequences  \\\n",
       "2315  ENST00000335295  ENSE00001829867  ...                      NaN   \n",
       "\n",
       "      Edit_in_Exon  Edit_in_CDS   GC%  # Edits/guide  Poly_T  \\\n",
       "2315          True         True  50.0              1   False   \n",
       "\n",
       "      mutation_on_guide  guide_change_mutation  mutation_on_window  \\\n",
       "2315               True                   True                True   \n",
       "\n",
       "      mutation_on_PAM  \n",
       "2315            False  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hbb_mut[hbb_mut.guide_change_mutation]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db8b2511-79b9-4ad2-8e4d-01667f708779",
   "metadata": {},
   "source": [
    "#### David Liu paper\n",
    "\n",
    "- WT codon is *GAG* and the mutant one is *GTG*\n",
    "- Mutation is on 5227002 and mutant sequence between 5227001-5227003 is CAC --> ABE --> CGC (+1 strand)\n",
    "\n",
    "GCG --> Ala --> naturally occurring, non-sickling variant hemoglobin \"Makassar\" (HbG) [https://doi.org/10.1182/blood-2020-139016]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "af56b91e-63f5-4520-9bea-a08739f7dfc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_server = \"https://www.ebi.ac.uk/proteins/api/\"\n",
    "uniprot_api = \"proteins?offset=0&size=-1&accession=D9YZU5\" \n",
    "api_request = requests.get(uniprot_server + uniprot_api, headers={\"Accept\": \"application/json\"})\n",
    "hbb_seq = None\n",
    "for i in api_request.json():\n",
    "    hbb_seq = i[\"sequence\"][\"sequence\"]\n",
    "\n",
    "hbb_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8641dc62-fe59-40d5-af32-11ec10074e86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'molecule': 'dna',\n",
       " 'query': '11:5226994-5227016',\n",
       " 'seq': 'ACTTCTCCTCAGGAGTCAGATGC',\n",
       " 'id': 'chromosome:GRCh38:11:5226994:5227016:1'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server = \"https://rest.ensembl.org\"\n",
    "hugo_ensembl = \"/sequence/region/human/11:5226994-5227016?content-type=text/plain\"\n",
    "gene_request = requests.get(server + hugo_ensembl, headers={\"Content-Type\": \"application/json\"})\n",
    "gene_request.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29cf53ea-a8f3-4643-b7aa-8db8869017da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe49c796-7609-4bdf-9827-acf40f84c099",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
